src/ui/analyse/forecast/ForecastCtrl.ts

```diff
@@ -0,0 +1,195 @@
+import { keyBy, pickBy, some, values } from 'lodash-es'
+import { AnalyseDataForForecast, ForecastStep, MinimalForecastStep } from '~/lichess/interfaces/forecast'
+import router from '~/router'
+import redraw from '~/utils/redraw'
+import { playAndSaveForecasts, saveForecasts } from './xhr'
+
+const MAX_FORECAST_PLIES = 30
+
+type SanMap = {[san: string]: ForecastStep[]}
```

Comment:
I'd use a `Map` here instead. This is a better data structure for this usage, and I think we can easily avoid all the lodash functions above